src/RamOperation.h

```diff
@@ -288,6 +288,145 @@ class RamIndexScan : public RamIndexRelationSearch {
+    std::unique_ptr<RamCondition> condition;
+};
+
+/**
+ * Relation Choice with Index
+ *
+ * Find a tuple in relation such that condition holds.
+ */
+class RamIndexChoice : public RamIndexRelationSearch {
+public:
+    RamIndexChoice(std::unique_ptr<RamRelationReference> r, int ident, std::unique_ptr<RamCondition> cond,
+            std::vector<std::unique_ptr<RamExpression>> queryPattern, std::unique_ptr<RamOperation> nested,
+            std::string profileText = "")
+            : RamIndexRelationSearch(std::move(r), ident, std::move(queryPattern), std::move(nested),
+                      std::move(profileText)),
+              condition(std::move(cond)) {
+        assert(getRangePattern().size() == getRelation().getArity());
+    }
+
+    /** get condition */
+    const RamCondition& getCondition() const {
+        return *condition;
+    }
+
+    /** Print */
+    void print(std::ostream& os, int tabpos) const override {
+        const RamRelation& rel = getRelation();
+        os << times(" ", tabpos);
+        os << "CHOICE " << rel.getName() << " AS t" << getTupleId() << " INDEX ";
+        bool first = true;
+        for (unsigned int i = 0; i < rel.getArity(); ++i) {
+            if (queryPattern[i] != nullptr) {
+                if (first) {
+                    first = false;
+                } else {
+                    os << " and ";
+                }
+                os << "t" << getTupleId() << "." << rel.getArg(i) << "=";
+                queryPattern[i]->print(os);
+            }
+        }
+        os << " WHERE " << getCondition();
+        os << std::endl;
+        RamIndexRelationSearch::print(os, tabpos + 1);
+    }
+
+    /** Apply mapper */
+    void apply(const RamNodeMapper& map) override {
+        RamRelationSearch::apply(map);
+        for (auto& cur : queryPattern) {
+            if (cur != nullptr) {
+                cur = map(std::move(cur));
+            }
+        }
+        condition = map(std::move(condition));
+    }
+
+    std::vector<const RamNode*> getChildNodes() const override {
+        auto res = RamIndexRelationSearch::getChildNodes();
+        if (condition != nullptr) {
+            res.push_back(condition.get());
+        }
+        return res;
+    }
+
+    RamIndexChoice* clone() const override {
+        std::vector<std::unique_ptr<RamExpression>> resQueryPattern(queryPattern.size());
+        for (unsigned int i = 0; i < queryPattern.size(); ++i) {
+            if (nullptr != queryPattern[i]) {
+                resQueryPattern[i] = std::unique_ptr<RamExpression>(queryPattern[i]->clone());
+            }
+        }
+        RamIndexChoice* res = new RamIndexChoice(std::unique_ptr<RamRelationReference>(relationRef->clone()),
+                getTupleId(), std::unique_ptr<RamCondition>(condition->clone()), std::move(resQueryPattern),
+                std::unique_ptr<RamOperation>(getOperation().clone()), getProfileText());
+        return res;
+    }
+
+protected:
+    /** Condition */
```

Comment:
This is not a helpful comment. Either be verbose enough for it to describe the point of the condition, or remove it.